### PR TITLE
backend: add service discovery interface and implement for single host deployments

### LINKED ISSUE #1847: RFC: Backend Service Discovery
# Background

Backend plugins (including the proxy and graphql plugins) that supply different pieces of functionality in the Backstage ecosystem need to communicate amongst each other. When they are deployed separately, they need to be able to locate the API endpoints of other plugins and services, in order to make requests to them.

Examples:

- The payload of a scaffolder request may contain the ID of a template, which the scaffolder wants to resolve in the catalog. It would then make an HTTP request to the catalog API for the data of that particular template.
- The graphql plugin may be deployed as part of the main backend, and its resolvers may then want to call out to a separately deployed catalog, lighthouse, etc.

# Proposed Interface

We propose to start out with a minimal, fairly explicit, code-based solution to start with. A plugin, when being initialized, is passed an instance of a service lookup interface that is able to find endpoints based on a plugin ID.

Example implementation:

```ts
/**
 * A basic service endpoint resolver.
 */
export type EndpointResolver = {
  /**
   * Returns the current set of endpoints where the given service can be reached.
   *
   * @param serviceName The opaque "name" to look up, for example a plugin ID
   * @returns A list of URLs to endpoint roots
   */
  resolve(serviceName: string): Promise<URL[]>;
}
```

A service name is an opaque identifier of the service whose endpoints we want to reach. We suggest that initially these are just plain backend plugin IDs, e.g. `"catalog"`.

The returned list consists of zero or more URLs. The caller may choose freely among them with any strategy. The endpoint resolver MAY choose to randomize the order of returned items on each call to support naive load balancing.

The `resolve` call is expected to be reasonably fast. The caller shall be able to call it for each outgoing request without worry for negative performance impact from the resolver itself.

# Initial Implementation

The first implementation of this interface may simply be driven by the app-config and/or have some hard coded behaviors. This can be extended later to more advanced resolvers. The company adopting Backstage will also be free to inject specific implementations of this interface that match their unique service lookup needs.

# Future Extension

Some ideas for future additions:

* Add SRV record lookup support
* Add support for bridging to third party service discovery services
* Add health checks and return known healthy endpoints
* Implement some load balancing scheme

# Limitations

This solution does not initially take health checking, fallbacks and load balancing into account. Those are hard to perform properly without being in the actual request path. When several endpoints are returned, it is up to clients to deal with that as they see fit.

### LINKED ISSUE #2596: Use DiscoveryApi to find the catalog API route in techdocs-backend router.

### PR DESCRIPTION
**Pull request by Rugvip**
Fixes #1847, #2596

Went with an interface similar to the frontend `DiscoveryApi`, since it's dead simple but still provides a lot of flexibility in the implementation.

Also ended up with two different methods, one for internal endpoint discovery and one for external. The two use-cases are explained a bit more in the docs, but basically it's service-to-service vs callback URLs.

This did get me thinking about uniqueness and that we're heading towards a global namespace for backend plugin IDs. That's probably fine tbh, but if we're happy with that we should leverage it a bit more to simplify the backend setup. For example we'd have each plugin provide its own ID and not manually mount on paths in the backend.

Draft until we're happy with the implementation, then I can add more docs and changelog entry. Also didn't go on a thorough hunt for places where discovery can be used, but tbh I don't think there are many since it's been pretty awkward to do service-to-service communication.

### CODE REVIEW

File: plugins/techdocs-backend/src/service/router.ts [hunk 1]
```diff
@@ -52,20 +56,25 @@ export async function createRouter({
...
-      await fetch(
-        `${baseUrl}/api/catalog/entities/by-name/${kind}/${namespace}/${name}`,
-      )
-    ).json()) as Entity;
+    const catalogUrl = await discovery.getBaseUrl('catalog');
+    const triple = [kind, namespace, name].map(encodeURIComponent).join('/');
+
+    const catalogRes = await fetch(`${catalogUrl}/entities/by-name/${triple}`);
+    if (!catalogRes.ok) {
+      catalogRes.body.pipe(res.status(catalogRes.status));
```
Comment (freben) [inline plugins/techdocs-backend/src/service/router.ts:73]: Fancy. And this works whether there's a body or not?

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  static fromConfig(config: Config, options?: { basePath?: string }) {
+    const basePath = options?.basePath ?? '/api';
+    const externalBaseUrl = config.getString('backend.baseUrl');
+
+    const {
+      listenHost = '127.0.0.1',
+      listenPort = DEFAULT_PORT,
+    } = readBaseOptions(config.getConfig('backend'));
+    const protocol = config.has('backend.https') ? 'https' : 'http';
+    const internalBaseUrl = `${protocol}://${listenHost}:${listenPort}`;
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-common/src/discovery/SingleHostDiscovery.ts]: That should work just fine, changed it to support IPv6 though

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { Config } from '@backstage/config';
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-common/src/discovery/SingleHostDiscovery.ts:1]: fkin auth import been screwed up lately :p

File: plugins/techdocs-backend/src/service/router.ts (hunk 1 above)
Comment (Rugvip) [reply plugins/techdocs-backend/src/service/router.ts:73]: Should, an empty pipe is still a pipe, no? :grin:

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/backend-common/src/discovery/SingleHostDiscovery.ts]: Probably nicer to translate those to localhost even it happens to work though, updated that too

File: packages/backend-common/src/discovery/SingleHostDiscovery.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      config.getConfig('backend'),
+    );
+    const protocol = config.has('backend.https') ? 'https' : 'http';
+
+    // Translate bind-all to localhost, and support IPv6
+    let host = listenHost;
+    if (host === '::') {
+      host = '::1';
+    } else if (host === '0.0.0.0') {
+      host = '127.0.0.1';
```
Comment (freben) [inline packages/backend-common/src/discovery/SingleHostDiscovery.ts:54]: Great, I was just about to comment on the previous version that we can't assume that a bind address is routable. Now, can both the above options result in just `localhost`, or is that more risky?

**Review by benjdlambert — APPROVED** [review]
Looks good to me. Any docs we should update too?

**Comment by Rugvip** [comment]
@benjdlambert couldn't find any docs to update, cept for the changelog

**Comment by benjdlambert** [comment]
Ok maybe one for an upcoming docs party.

**Review by freben — APPROVED** [review]

**Comment by benjdlambert** [comment]
@spotify/techdocs-core can I get your approval too?

**Review by hooloovooo — APPROVED** [review]

**Pull request merged**